daft/execution/physical_plan.py

```diff
@@ -564,6 +564,78 @@ def broadcast_join(
                 return
 
 
+def cross_join(
+    left_plan: InProgressPhysicalPlan[PartitionT],
+    right_plan: InProgressPhysicalPlan[PartitionT],
+    outer_loop_side: JoinSide,
+):
+    stage_id = next(stage_id_counter)
+
+    outer_plan, inner_plan = (left_plan, right_plan) if outer_loop_side == JoinSide.Left else (right_plan, left_plan)
+
+    inner_requests: deque[SingleOutputPartitionTask[PartitionT]] = deque()
+    for step in inner_plan:
+        if isinstance(step, PartitionTaskBuilder):
+            step = step.finalize_partition_task_single_output(stage_id=stage_id)
+            inner_requests.append(step)
+        yield step
+
+    outer_requests: deque[SingleOutputPartitionTask[PartitionT]] = deque()
+    while True:
+        while outer_requests and outer_requests[0].done():
+            next_outer = outer_requests.popleft()
+
+            for next_inner in inner_requests:
+                while not next_inner.done():
+                    logger.debug(
+                        "cross join blocked on completion of inner side of join.\n inner sources: %s",
+                        inner_requests,
+                    )
+                    yield None
+
+                next_left, next_right = (
+                    (next_outer, next_inner) if outer_loop_side == JoinSide.Left else (next_inner, next_outer)
+                )
+
+                # Calculate memory request for task.
+                left_size_bytes = next_left.partition_metadata().size_bytes
+                right_size_bytes = next_right.partition_metadata().size_bytes
+                if left_size_bytes is None and right_size_bytes is None:
+                    size_bytes = None
+                elif left_size_bytes is None and right_size_bytes is not None:
+                    # Assume that left and right side are ~ the same size.
+                    size_bytes = right_size_bytes**2
+                elif right_size_bytes is None and left_size_bytes is not None:
+                    # Assume that left and right side are ~ the same size.
+                    size_bytes = left_size_bytes**2
+                elif left_size_bytes is not None and right_size_bytes is not None:
+                    size_bytes = left_size_bytes * right_size_bytes
+
+                join_step = PartitionTaskBuilder[PartitionT](
+                    inputs=[next_left.partition(), next_right.partition()],
+                    partial_metadatas=[next_left.partition_metadata(), next_right.partition_metadata()],
+                    resource_request=ResourceRequest(memory_bytes=size_bytes),
+                ).add_instruction(instruction=execution_step.CrossJoin(outer_loop_side=outer_loop_side))
+
+                yield join_step
+
+        try:
+            step = next(outer_plan)
```

Comment:
this generator looks correct to me. However one issue is the the outer side doesn't spawn a task until the inner side is all done. This will cause a stall especially when operating on a larger cluster

imagine we have 2 left partitions and 3 right. and that the left is the other.
current task order:
```
[R1, R2, R3], [L1], [L1R1, L1R2, L1R2], [L2], ...
```
We have limited parallelism when it comes to materializing the left side.

instead you could do
```
[R1, R2, R3, L1], [L1R1, L1R2, L1R2, L2],

```